rest/route/distro.go

```diff
@@ -583,62 +584,179 @@ func validateDistro(ctx context.Context, apiDistro *model.APIDistro, resourceID
+			if err := h.env.RemoteQueue().Put(ctx, units.NewHostExecuteJob(h.env, host, h.opts.Script, h.opts.Sudo, h.opts.SudoUser, ts)); err != nil {
+				catcher.Wrapf(err, "problem enqueueing job to run script on host '%s'", host.Id)
+				continue
+			}
+			hostIDs = append(hostIDs, host.Id)
+		}
 	}
 	if catcher.HasErrors() {
-		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "problem enqueueing jobs to run script on hosts"))
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrapf(err, "problem enqueueing jobs to run script on hosts"))
 	}
 
-	return gimlet.NewJSONResponse(struct{}{})
+	return gimlet.NewJSONResponse(struct {
+		HostIDs []string `json:"host_ids"`
+	}{HostIDs: hostIDs})
+}
+
+///////////////////////////////////////////////////////////////////////
+//
+// PATCH /rest/v2/distros/{distro}/icecream_config
+
+type distroIcecreamConfigHandler struct {
+	distro string
+	opts   model.APIDistroScriptOptions
+	sc     data.Connector
+	env    evergreen.Environment
+}
+
+func makeDistroIcecreamConfig(sc data.Connector, env evergreen.Environment) gimlet.RouteHandler {
+	return &distroIcecreamConfigHandler{
+		sc:  sc,
+		env: env,
+	}
+}
+
+func (h *distroIcecreamConfigHandler) Factory() gimlet.RouteHandler {
+	return &distroIcecreamConfigHandler{
+		sc:  h.sc,
+		env: h.env,
+	}
+}
+
+// Parse extracts the distro and JSON payload from the http request.
+func (h *distroIcecreamConfigHandler) Parse(ctx context.Context, r *http.Request) error {
+	h.distro = gimlet.GetVars(r)["distro_id"]
+	body := util.NewRequestReader(r)
+	defer body.Close()
+
+	if err := util.ReadJSONInto(body, &h.opts); err != nil {
+		return errors.Wrap(err, "could not read request body")
+	}
+
+	return nil
+}
+
+// Run enqueues a job to run a script on all hosts that are not down for the
+// given given distro ID.
+func (h *distroIcecreamConfigHandler) Run(ctx context.Context) gimlet.Responder {
+	hosts, err := h.sc.FindHostsByDistro(h.distro)
+	if err != nil {
+		return gimlet.MakeJSONErrorResponder(errors.Wrapf(err, "could not find hosts for the distro '%s'", h.distro))
+	}
+
+	dat, err := distro.NewDistroAliasesLookupTable()
+	if err != nil {
+		return gimlet.MakeJSONInternalErrorResponder(errors.Wrap(err, "could not get distro lookup table"))
+	}
+
+	catcher := grip.NewBasicCatcher()
+	var hostIDs []string
+	for _, host := range hosts {
+		// If the distro exists, we use the settings directly from that distro;
+		// if the distro in the host document is deleted, we make a best-effort
+		// attempt to resolve it to a real distro by attempting to pick any
+		// existing distro with an alias that matches the deleted distro.
+		distroIDs := dat.Expand([]string{host.Distro.Id})
+		if len(distroIDs) == 0 {
+			catcher.Errorf("could not look up distro '%s'", host.Distro.Id)
+			continue
+		}
+		var d distro.Distro
+		var distroFound bool
+		for _, distroID := range distroIDs {
```

Comment:
is there a form of this query that does a $in?

Reply:
Changed to use `distro.ByIds()`